### PR TITLE
changes for text detection

### PR DESCRIPTION
**Pull request by GalyaZalesskaya**

### CODE REVIEW

File: mmdet/core/evaluation/coco_utils.py [hunk 1]
```diff
@@ -50,6 +50,14 @@ def coco_eval(result_files,
         cocoEval.accumulate()
         cocoEval.summarize()
 
+        predictions = cocoEval.cocoDt.imgToAnns
```
Comment (Ilya-Krylov) [inline mmdet/core/evaluation/coco_utils.py]: probably it is worth using `result_types` to enable text detection metrics computation.
In common case we don't need to perform text detection quality estimation

**Comment by Ilya-Krylov** [comment]
@GalyaZalesskaya please press "Ready for review" as soon as you ready for review. Don't forget to rebase on the latest ote branch state

**Comment by Ilya-Krylov** [comment]
please change target branch to develop

**Comment by GalyaZalesskaya** [comment]
@Ilya-Krylov @druzhkov-paul Hello, I've finished with all my work and checks. Please, review.

**Comment by GalyaZalesskaya** [comment]
@Ilya-Krylov I've changed base branch to "develop" like you did [here](https://github.com/opencv/mmdetection/pull/44) and rebased on it. Is it right?

File: mmdet/datasets/coco.py [hunk 2]
```diff
@@ -384,18 +386,19 @@ def evaluate(self,
...
                 continue
 
-            if metric not in result_files:
-                raise KeyError(f'{metric} is not in results')
+            metric_type = 'bbox' if metric == 'f1' else metric
+            if metric_type not in result_files:
+                raise KeyError(f'{metric_type} is not in results')
             try:
-                cocoDt = cocoGt.loadRes(result_files[metric])
+                cocoDt = cocoGt.loadRes(result_files[metric_type])
```
Comment (Ilya-Krylov) [inline mmdet/datasets/coco.py:393]: why did you rename `metric` to `metric_type`?

Comment (GalyaZalesskaya) [reply mmdet/datasets/coco.py:393]: To evaluate f1, I  need to load 'bbox' results. So, metric_type = 'bbox' if the we calculate 'f1' and, otherwise, metric_type=metric, like it was before my changes.
```metric_type = 'bbox' if metric == 'f1' else metric```  (389 line)

**Comment by Ilya-Krylov** [comment]
@druzhkov-paul could you have a look?

File: mmdet/datasets/coco.py [hunk 3]
```diff
@@ -384,18 +386,19 @@ def evaluate(self,
...
             except IndexError:
                 print_log(
                     'The testing results of the whole dataset is empty.',
                     logger=logger,
                     level=logging.ERROR)
                 break
 
-            iou_type = 'bbox' if metric == 'proposal' else metric
+            iou_type = 'bbox' if metric == 'proposal' or 'f1' else metric
```
Comment (druzhkov-paul) [inline mmdet/datasets/coco.py]: What's happening here?

File: mmdet/datasets/coco.py [hunk 4]
```diff
@@ -416,6 +419,17 @@ def evaluate(self,
...
+
+                if metric == 'f1':
+                    predictions = cocoEval.cocoDt.imgToAnns
+                    gt_annotations = cocoEval.cocoGt.imgToAnns
+                    recall, precision, hmean, _ = text_eval(
+                        predictions, gt_annotations, test_cfg.score_thr,
+                        show_recall_graph=False,
+                        use_transcriptions=False)
+                    print('Text detection recall={:.4f} precision={:.4f} hmean={:.4f}'.
+                          format(recall, precision, hmean))
```
Comment (druzhkov-paul) [inline mmdet/datasets/coco.py:428]: Are the results returned somehow or just printed?

File: mmdet/datasets/coco.py [hunk 5]
```diff
@@ -384,18 +386,19 @@ def evaluate(self,
                 print_log(log_msg, logger=logger)
                 continue
 
-            if metric not in result_files:
-                raise KeyError(f'{metric} is not in results')
+            metric_type = 'bbox' if metric == 'f1' else metric
+            if metric_type not in result_files:
+                raise KeyError(f'{metric_type} is not in results')
```
Comment (druzhkov-paul) [inline mmdet/datasets/coco.py:391]: Does it mean that test script must be run with `bbox` evaluation type specified? Can you please provide an example of how to run testing script right to get an `f1` metric value as an output?

**Comment by GalyaZalesskaya** [comment]
run tests

File: mmdet/datasets/coco.py [hunk 6]
```diff
@@ -416,6 +419,17 @@ def evaluate(self,
                 cocoEval.evaluate()
                 cocoEval.accumulate()
                 cocoEval.summarize()
+
```
Comment (druzhkov-paul) [inline mmdet/datasets/coco.py:419]: Does `text_eval` rely on `cocoEval.evaluate()` results? If no, why do you need to call it all the time?

File: mmdet/datasets/coco.py (hunk 4 above)
Comment (GalyaZalesskaya) [reply mmdet/datasets/coco.py:428]: Just printed. I did the output in the same manner as bbox calculating. There results are printed only, while intermediate results from each iteration (loss, loss_cls, loss_bbox etc) are going to the log file.

File: mmdet/datasets/coco.py (hunk 6 above)
Comment (GalyaZalesskaya) [reply mmdet/datasets/coco.py:419]: You are right. Will fix it

File: mmdet/datasets/coco.py (hunk 3 above)
Comment (GalyaZalesskaya) [reply mmdet/datasets/coco.py]: There is a line in pycocotools/cocoeval.py
  ```  #  iouType    - ['segm'] set iouType to 'segm', 'bbox' or 'keypoints' ``` (32 line)
And to calculate f1-score I use bbox type of layout, that I've defined here

File: mmdet/datasets/coco.py (hunk 5 above)
Comment (GalyaZalesskaya) [reply mmdet/datasets/coco.py:391]:
```
python test.py $CONFIG $SNAPSHOT
--eval f1
--out result.pkl
```
or 
```
python test_exported.py $CONFIG $SNAPSHOT
--json_out result
--eval f1
```
The output will be like 
```
Text detection recall=0.8639 precision=0.9061 hmean=0.8845
```

File: mmdet/datasets/coco.py (hunk 3 above)
Comment (druzhkov-paul) [reply mmdet/datasets/coco.py]: This is clear. But, can you please explain me only this line of code.

File: mmdet/datasets/coco.py (hunk 4 above)
Comment (druzhkov-paul) [reply mmdet/datasets/coco.py:428]: Is it Ok to just print the result?

@Ilya-Krylov what do you think?

File: mmdet/datasets/coco.py (hunk 5 above)
Comment (druzhkov-paul) [reply mmdet/datasets/coco.py:391]: Ah, I see. `bbox` will be in `result_files` after running `results2json` independently on whether metric is `f1` or not (`proposal` is the only exception).

File: mmdet/datasets/coco.py (hunk 4 above)
Comment (druzhkov-paul) [reply mmdet/datasets/coco.py:428]: BTW, is it possible to calculate this F1 score during training time evaluation?

File: mmdet/datasets/coco.py (hunk 3 above)
Comment (GalyaZalesskaya) [reply mmdet/datasets/coco.py]: Iou_type is `bbox` if the metric is `bbox`, `proposal` or `f1`. In the other way, if metric is `segm`, iou_type will be `segm`.
Do you suggest to rewrite this line? Something like that
```
iou_type = 'segm' if metric == 'segm' else 'bbox'
```

File: mmdet/datasets/coco.py (hunk 4 above)
Comment (GalyaZalesskaya) [reply mmdet/datasets/coco.py:428]: I thought about it, but to calculate it, we should put f1 eval parameter somewhere in the train.py call, like we do it in test.py, using `--eval f1`. 
So it requires some changes in the logic of the code.

File: mmdet/datasets/coco.py (hunk 3 above)
Comment (druzhkov-paul) [reply mmdet/datasets/coco.py]: And now please evaluate this expression with `metric == 'segm'`.

Comment (druzhkov-paul) [inline mmdet/datasets/coco.py]:
```suggestion
            iou_type = 'bbox' if metric in {'proposal', 'f1'} else metric
```

Comment (GalyaZalesskaya) [reply mmdet/datasets/coco.py]: Got it! Thanks you very much!

File: mmdet/datasets/coco.py (hunk 4 above)
Comment (druzhkov-paul) [reply mmdet/datasets/coco.py:428]: All parameters are passed via config, aren't they?

File: mmdet/datasets/coco.py (hunk 3 above)
Comment (GalyaZalesskaya) [reply mmdet/datasets/coco.py:401]: Is there some secret reason why do you prefer dictionary instead  of list?

File: mmdet/datasets/coco.py (hunk 4 above)
Comment (GalyaZalesskaya) [reply mmdet/datasets/coco.py:428]: They are, but I still will have to add the parameter in the config. I can do it, but I need an approval from @Ilya-Krylov

File: mmdet/datasets/coco.py (hunk 3 above)
Comment (druzhkov-paul) [reply mmdet/datasets/coco.py]: It's a set, not a dict. Sets and list has different complexity of element search operation. No difference here, though.

Comment (Ilya-Krylov) [reply mmdet/datasets/coco.py]: Let's add metrics value to `eval_results`. Regarding evaluation during training Galya can add it later, BTW it does not require my approval. There is a need to have Galya's detector in OTE as soon as possible.

File: mmdet/datasets/coco.py (hunk 4 above)
Comment (Ilya-Krylov) [reply mmdet/datasets/coco.py:428]: Let's add metrics value to `eval_results`. Regarding evaluation during training  Galya can add it later. There is a need to have Galya's detector in OTE as soon as possible. That's why I propose to postpone this feature though it is quite useful.

Comment (druzhkov-paul) [reply mmdet/datasets/coco.py:428]: That's the thing you should've probably started with.  😉 Let's go that way.

IMO, adding anything to `eval_results` makes sense only for training time validation.

**Comment by GalyaZalesskaya** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
You need to add Polygon3 to runtime requirements

**Comment by GalyaZalesskaya** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
@druzhkov-paul all issues resolved? Do you approve changes?

**Review by druzhkov-paul — APPROVED** [review]
